### PR TITLE
Map Block: Use the block styles API for choosing map themes

### LINKED ISSUE #14559: Map Block: UI for map themes
@shaunandrews has pointed out the UI for the map themes would probably better to follow the block style pattern. https://github.com/Automattic/Dotcom-roadmap/issues/74#issuecomment-574770538

+1!

### PR DESCRIPTION
**Pull request by pento**
This PR switches the map block over to using the block styles attribute, rather than the custom theme picker.

Fixes #14559

#### Changes proposed in this Pull Request:
* See discussion on #14559 and linked issues.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Change to existing feature.

#### Testing instructions:

* Add a new map block, test that the block style switches the map theme as expected.
* Open an post with an old map block, check that in correctly converts to the new style.

#### Proposed changelog entry for your changes:
* Map Block: Improve the map theme picker.

### CODE REVIEW

**Comment by pento** [comment]
Some early notes:

- The Styles panel defaults to being closed, [there's no way to override that](https://github.com/WordPress/gutenberg/blob/5aaf132172049aabba4c72f1e7b088517bc0dd04/packages/block-editor/src/components/block-inspector/index.js#L67). This is a change in behaviour from the existing theme picker.
- I couldn't figure out the purpose of the "Default Style" dropdown which appears in the Style panel. Also not sure how to disable it.
- I haven't tested frontend rendering, either of old or new map blocks. They're almost certainly broken at the moment. 🙂

**Comment by scruffian** [comment]
> The Styles panel defaults to being closed, there's no way to override that. This is a change in behaviour from the existing theme picker.

Sounds like we should open a core patch for that

**Comment by scruffian** [comment]
> I haven't tested frontend rendering, either of old or new map blocks. They're almost certainly broken at the moment. 🙂

I did work before this change, but now it doesn't :)

**Review by scruffian — DISMISSED** [review]
This is a really good direction. We have [a custom component](https://github.com/Automattic/jetpack/blob/ef97833724912a78d17d9f5b0cf094c50a582259/extensions/shared/components/block-styles-selector/index.js) which does something similar, but if we can do it this way that would be ideal!

We just need to get it working on the frontend...

**Comment by scruffian** [comment]
There's an annoying margin at the top of the previews, which I have a fix for here: https://github.com/WordPress/gutenberg/pull/19983

File: extensions/blocks/map/edit.js
Comment (pablinos) [inline extensions/blocks/map/edit.js]: This construct was a bit tricky for me to understand. Is essentially doing this?

`const mapStyle = settings.styles.find( styleObject => className.includes( styleObect.name ) )`

File: extensions/blocks/map/edit.js
Comment (pablinos) [inline extensions/blocks/map/edit.js]: We seem to only use `mapStyle` as a string, so we could we defaulth this when it's set and leave this line as is?

File: extensions/blocks/map/index.js
Comment (pablinos) [inline extensions/blocks/map/index.js]: I may have missed it in the wider code, but I can't see where this attribute is getting set. As it is based on `className`, could it be derived in the `save` function too?

**Review by pablinos — COMMENTED** [review]
I agree with @scruffian, this is a better pattern than we have at the moment. Let's get it working along with any core patches needed!

**Comment by apeatling** [comment]
Ran through and tested this and it worked as expected and mentioned in the instructions. I like the experience of using block styles for this. Agreed that we should submit a core PR to allow the style panel to be open by default, it'll be incredibly easy to miss this otherwise.

File: extensions/blocks/map/edit.js
Comment (pento) [reply extensions/blocks/map/edit.js]: I've done the same thing in multiple different languages this week: using `reduce()` when I should've used `find()`. 🙃 Fixed in 9c35f1d86.

File: extensions/blocks/map/edit.js
Comment (pento) [reply extensions/blocks/map/edit.js]: `mapStyle.preview` is used for the `placholderPreview`. That said, I think I'm going to redo how it's storing the style, so it works the same way as prior to this PR. We can just ignore the custom class that Core will add to the block.

Comment (pento) [reply extensions/blocks/map/edit.js]: Turned out to be relatively easy, done in e2927a623.

**Comment by pento** [comment]
This PR has generated two Gutenberg PRs:

- As several folks suggested, something to keep the Styles panel open by default: WordPress/gutenberg#20617
- A block setting to remove the Default Style selector, since it's not really relevant to this block, and also kind of confusing: WordPress/gutenberg#20620

I've also switched back to storing the map style in the `data-map-style` attribute (e2927a6), which has the added bonus of no longer needing a deprecated block definition, and we're not changing the `save()` output, so we don't have to hold merging this to WPCOM until the end of the month. 🙂

File: extensions/blocks/map/edit.js
Comment (scruffian) [inline extensions/blocks/map/edit.js]: This is giving me an error: `TypeError: Cannot read property 'name' of undefined`.

I wonder if we should be using `import { getActiveStyleName } from '../../shared/block-styles';`

**Comment by scruffian** [comment]
I added a commit to bring this into line with how we do things with Tiled Gallery. It also makes the code a bit simpler.

**Review by Copons — DISMISSED** [review]
LGTM!

**Comment by pento** [comment]
Thank you for the suggestion, @scruffian!

I've updated this PR to match the changes in WordPress/gutenberg#20620, let's leave it sit for a bit until the Core behaviour is locked in.

**Comment by scruffian** [comment]
I'm not convinced we need to wait for WordPress/gutenberg#20620 to merge before we merge this. We can just go ahead without it and update when we know how that pans out...

**Review by jeherve — CHANGES_REQUESTED** [review]
I tested on WP 5.3, without the Gutenberg plugin.

Existing blocks seem to invalidate when switching to this branch. Here is an example of a block that was invalidated when I tried to edit it again:

```
<!-- wp:jetpack/map {"align":"wide","points":[{"placeTitle":"Rennes","title":"Rennes","caption":"Rennes, Ille-et-Vilaine, France","id":"place.7688731914683530","coordinates":{"longitude":-1.68083,"latitude":48.11417}}],"mapStyle":"terrain","mapDetails":false,"mapCenter":{"lng":-1.68083,"lat":48.11417},"scrollToZoom":true,"showFullscreenButton":false} -->
<div class="wp-block-jetpack-map alignwide" data-map-style="terrain" data-map-details="false" data-points="[{&quot;placeTitle&quot;:&quot;Rennes&quot;,&quot;title&quot;:&quot;Rennes&quot;,&quot;caption&quot;:&quot;Rennes, Ille-et-Vilaine, France&quot;,&quot;id&quot;:&quot;place.7688731914683530&quot;,&quot;coordinates&quot;:{&quot;longitude&quot;:-1.68083,&quot;latitude&quot;:48.11417}}]" data-zoom="13" data-map-center="{&quot;lng&quot;:-1.68083,&quot;lat&quot;:48.11417}" data-marker-color="red" data-scroll-to-zoom="true"><ul><li><a href="https://www.google.com/maps/search/?api=1&amp;query=48.11417,-1.68083">Rennes</a></li></ul></div>
<!-- /wp:jetpack/map -->
```

Coud you take a look?

**Comment by scruffian** [comment]
I added a commit to fix that!

**Review by jeherve — CHANGES_REQUESTED** [review]
Do you think you could move the deprecated save to a `deprecated/v1` to remain consistent with our other blocks like VideoPress or Tiled Galleries?

**Comment by scruffian** [comment]
> Do you think you could move the deprecated save to a deprecated/v1 to remain consistent with our other blocks like VideoPress or Tiled Galleries?

Done in b97e7ee

File: extensions/blocks/map/deprecated/v1/index.js
Comment (Copons) [inline extensions/blocks/map/deprecated/v1/index.js]: This way we are losing the `mapStyle` choice when migrating a block, reverting it to the default style.

To repro, try adding a Map block with style "satellite" on master, then checkout this PR, reload the post and observe how the map is now with style "default".

One way to fix this might be to convert `mapStyle` to the `is-style-` class:

```es6
migrate: oldAttributes => ( {
	...omit( oldAttributes, 'mapStyle' ),
	className: `${ oldAttributes.className } is-style-${ oldAttributes.mapStyle }`,
} ),
```

**Review by Copons — CHANGES_REQUESTED** [review]

File: extensions/blocks/map/deprecated/v1/index.js
Comment (pento) [reply extensions/blocks/map/deprecated/v1/index.js]: 👍🏻 1253f52ae

**Comment by pento** [comment]
I've updated the PR to address the remaining feedback.

While I was testing, it was interesting to note that we could keep the `mapStyle` block attribute as a write-only attribute, which would then allow us to merge this change to WPCOM immediately, rather than needing to wait for the next Jetpack release. If the block was created on a Simple site, then moved to Atomic, it would just read the `mapStyle` attribute, instead.

Given the proximity of the upcoming release, I don't think it's an issue to just wait until then to merge it.

File: extensions/blocks/map/deprecated/v1/index.js
Comment (Copons) [inline extensions/blocks/map/deprecated/v1/index.js]: Is this because blocks are not supposed to have more than one `is-style-` class?

**Review by Copons — DISMISSED** [review]
This seems to work fine for me! ✨ 

> Given the proximity of the upcoming release, I don't think it's an issue to just wait until then to merge it.

I agree, there's no big rush to merge this!

File: extensions/blocks/map/deprecated/v1/index.js
Comment (jeherve) [inline extensions/blocks/map/deprecated/v1/index.js]: I am getting a blank screen when trying to edit an old post that has 2 maps with custom styles, once I have switched to this branch:

image

It seems to happen here during the migration.

**Review by jeherve — DISMISSED** [review]

File: extensions/blocks/map/deprecated/v1/index.js
Comment (pento) [reply extensions/blocks/map/deprecated/v1/index.js]: Yup. It's not entirely self explanatory, so I've added a comment to explain what's going on.

Comment (pento) [reply extensions/blocks/map/deprecated/v1/index.js]: Fixed in 285323a0d.

**Review by Copons — APPROVED** [review]
Hopefully for the last time 😄

**Review by kraftbj — APPROVED** [review]
I tested it on a fresh install and on an install bringing in an older version of the block. Both worked as expected as described in the ticket.

Fusion is failing because of new files. I'm working on that now to update the phab.

**Comment by pento** [comment]
@Automattic/jetpack-crew: Could someone merge this, please? It looks like the tests are now passing on the diff (thanks, @kraftbj!), but it hasn't synced the status, so merge is blocked for me.

**Comment by kraftbj** [comment]
@pento Unblocked! Please merge this (and the corresponding WP.com diff) at your leisure.

**Pull request merged**
